### PR TITLE
[Role] az role assignment create/update: Support `--description`, `--condition` and `--condition-version`

### PR DESCRIPTION
**Pull request by jiasli**
Close https://github.com/Azure/azure-cli/issues/14552

### Design

New arguments to add to the existing arguments:

```
Arguments
    --description             : Description of the role assignment.
    --condition               : Condition under which the user can be granted permission
    --condition-version       : Version of the condition syntax
```

### Examples

```sh
# line breaks for legibility
az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com
                          --role reader
                          --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1
                          --description "Role assignment foo to check on bar"
                          --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'"
                          --condition-version "2.0"
```

Sample Output:

```json
{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

#### Description

* `description` is a string.
* If `description` is empty, replace with null.

#### Condition

* `condition` is a string.
* If `condition` is empty replace with null.

#### Condition-Version

* `condition-version` is a string.
* `condition-version` must be > `1.0`.
  - The service already has this check: `The specified role assignment ConditionVersion '0.1' is not supported`.
* If `condition-version` is empty and `condition` is set, `condition-version` defaults to `2.0`.
* If `condition-version` is set, `condition` must be set as well.

### Design

``` 
Command
    az role assignment update : update an existing role assignment for a user, group, or service
    principal.

Arguments
    --role-assignment   [Required] :  Description of a role as JSON, or a path to a file containing a JSON description.
```

### Example

```sh
az role assignment update --role-assignment '{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals '"'"'foo'"'"'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}'

# or
az role assignment update --role-assignment "C:\Path\to\JSON\roleAssignment.json
```

Sample Output:

```json
{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

### Checks

Same as for `role assignment create` with some small exceptions/additions:

* Only `description`, `condition` and `conditionVersion` can be modified.
  - This requires lots of efforts for diffing the original REST response and user input and should be implemented on the service side. NOT IMPLEMENTED in CLI yet.
* if no role assignment exists with the indicated properties, throw an exception indicating as such.
* `conditionVersion` cannot be downgraded to `1.X`.

## Testing Guide

```sh
az group create -l westus -n rg1

az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com --role reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1 --description "Role assignment foo to check on bar" --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'" --condition-version "2.0"

az role assignment update --role-assignment '{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals '"'"'foo'"'"'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/9e8947f9-a813-4003-bbdd-98fa57d4dca0",
  "name": "9e8947f9-a813-4003-bbdd-98fa57d4dca0",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalName": "admin4@AzureSDKTeam.onmicrosoft.com",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "roleDefinitionName": "Reader",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}'

az role assignment delete -g rg1

az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com --role reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1 --description "Role assignment foo to check on bar" --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'" --condition-version "2.0" > output.json

az role assignment update --role-assignment output.json

az group delete -n rg1
```

### CODE REVIEW

**Comment by jiasli** [comment]
This PR includes fixes for command bug #14461 and all live tests for `role` module #14317.

**Comment by yonzhan** [comment]
Role

File: src/azure-cli/azure/cli/command_modules/role/_help.py [hunk 1]
```diff
@@ -585,12 +585,43 @@
...
+helps['role assignment update'] = """
+type: command
+short-summary: Update an existing role assignment for a user, group, or service principal.
+examples:
+  - name: Update a role assignment from a JSON file.
+    text: az role assignment update --role-assignment assignment.json
+  - name: Update a role assignment from a JSON string. (Bash)
+    text: |
+        az role assignment update --role-assignment '{
+            "canDelegate": null,
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/role/_help.py:610]: Can we add parameters like --can-deleagate so that users can see help messages.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/_help.py:610]: Taking JSON as an input is asked by the service team. I had the same suggestion as you to expose explicit arguments but the suggestion was rejected.

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/role/_help.py:610]: CLI is yours. You fix it. Be brave. Go ahead.

File: src/azure-cli/azure/cli/command_modules/role/_help.py [hunk 2]
```diff
@@ -585,12 +585,43 @@
 examples:
   - name: Create role assignment for an assignee.
     text: az role assignment create --assignee sp_name --role a_role
+  - name: Create role assignment for an assignee with description and condition. (Only for 2020-04-01-preview API version and later.)
+    text: >-
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/role/_help.py:589]: What's the difference between `>-` and `|`?

**Review by qwordy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 3]
```diff
@@ -129,25 +130,38 @@ def _search_role_definitions(cli_ctx, definitions_client, name, scopes, custom_r
...
+    """Check parameters are provided correctly, then call _create_role_assignment."""
     if bool(assignee) == bool(assignee_object_id):
         raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
 
     if assignee_principal_type and not assignee_object_id:
         raise CLIError('usage error: --assignee-object-id GUID [--assignee-principal-type]')
 
+    # If condition is set and condition-version is empty, condition-version defaults to "2.0".
+    if condition and not condition_version:
+        condition_version = "2.0"
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:144]: How about adding this default value in the help message ?

File: src/azure-cli/azure/cli/command_modules/role/_help.py (hunk 2 above)
Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/role/_help.py:589]: @qwordy 
image
In short: `>-` will replace all single `\n` to whitespace, and combine double `\n` to one.

**Review by bim-msft — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/role/custom.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:144]: Nice catch. Added.

File: src/azure-cli/azure/cli/command_modules/role/_params.py [hunk 4]
```diff
@@ -186,6 +186,9 @@ def load_arguments(self, _):
                    "use the object id and not the app id.")
         c.argument('ids', nargs='+', help='space-separated role assignment ids')
         c.argument('include_classic_administrators', arg_type=get_three_state_flag(), help='list default role assignments for subscription classic administrators, aka co-admins')
+        c.argument('description', min_api='2020-04-01-preview', help='Description of role assignment.')
+        c.argument('condition', min_api='2020-04-01-preview', help='Condition under which the user can be granted permission.')
+        c.argument('condition_version', min_api='2020-04-01-preview', help='Version of the condition syntax. If --condition is specified without --condition-version, default to 2.0.')
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/_params.py:191]: > How about adding this default value in the help message ?

Added.

File: src/azure-cli/azure/cli/command_modules/role/_help.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/_help.py:589]: I use `>-` to make the code more readable purposefully.

**Review by jsntcy — APPROVED** [review]
:shipit:

**Review by arrownj — APPROVED** [review]

**Pull request merged**